packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -109,73 +132,22 @@ export class DefaultCompletionProvider implements CompletionProvider {
         this.grammar = services.Grammar;
         this.completionParser = services.parser.CompletionParser;
         this.nameProvider = services.references.NameProvider;
-        this.grammarConfig = services.parser.GrammarConfig;
+        this.lexer = services.parser.Lexer;
         this.nodeKindProvider = services.shared.lsp.NodeKindProvider;
         this.fuzzyMatcher = services.shared.lsp.FuzzyMatcher;
     }
 
     async getCompletion(document: LangiumDocument, params: CompletionParams): Promise<CompletionList | undefined> {
-        const root = document.parseResult.value;
-        const cst = root.$cstNode;
-        if (!cst) {
-            return undefined;
-        }
         const items: CompletionItem[] = [];
-        const textDocument = document.textDocument;
-        const text = textDocument.getText();
-        const offset = textDocument.offsetAt(params.position);
-        const acceptor: CompletionAcceptor = value => {
-            const completionItem = this.fillCompletionItem(textDocument, offset, value);
+        const contexts = this.buildContexts(document, params.position);
```

Comment:
Can we turn `items` to a Stream?

Reply:
What for? We need to constantly append to it, and once we're done, we need to convert it to an array for the LSP response.

Reply:
It didn't really make sense as a stream, but I've changed `buildContexts` into a generator function. That way, if we end early, we don't need to compute the other contexts.